rclpy/rclpy/executors.py

```diff
@@ -292,73 +277,50 @@ def wait_for_ready_callbacks(self, timeout_sec=None, nodes=None):
                 timers.append(timeout_timer)
 
             # Construct a wait set
-            with _WaitSet() as wait_set:
-                _rclpy.rclpy_wait_set_init(
-                    wait_set,
-                    len(subscriptions),
-                    num_guard_conditions,
-                    len(timers),
-                    len(clients),
-                    len(services))
-
-                entities = {
-                    'subscription': (subscriptions, 'subscription_handle'),
-                    'client': (clients, 'client_handle'),
-                    'service': (services, 'service_handle'),
-                    'timer': (timers, 'timer_handle'),
-                }
-                for entity, (handles, handle_name) in entities.items():
-                    _rclpy.rclpy_wait_set_clear_entities(entity, wait_set)
-                    for h in handles:
-                        _rclpy.rclpy_wait_set_add_entity(
-                            entity, wait_set, h.__getattribute__(handle_name)
-                        )
-                _rclpy.rclpy_wait_set_clear_entities('guard_condition', wait_set)
-                _rclpy.rclpy_wait_set_add_entity('guard_condition', wait_set, sigint_gc)
-                _rclpy.rclpy_wait_set_add_entity(
-                    'guard_condition', wait_set, self._guard_condition)
-
-                # Wait for something to become ready
-                _rclpy.rclpy_wait(wait_set, timeout_nsec)
-
-                # get ready entities
-                subs_ready = _rclpy.rclpy_get_ready_entities('subscription', wait_set)
-                guards_ready = _rclpy.rclpy_get_ready_entities('guard_condition', wait_set)
-                timers_ready = _rclpy.rclpy_get_ready_entities('timer', wait_set)
-                clients_ready = _rclpy.rclpy_get_ready_entities('client', wait_set)
-                services_ready = _rclpy.rclpy_get_ready_entities('service', wait_set)
+            wait_set = _WaitSet()
```

Comment:
This can't rely on `__del__` being called to destroy the allocated C wait set. That might just never happen. Instead the code should explicitly call a cleanup function when it doesn't need `wait_set` anymore.

Reply:
AFAIK an explicit method cannot be relied on unless it is implemented using the `with` or `try/finally` statements. I can add `__enter__()` and `__exit__()` methods.

I think `__del__` is OK here unless the C structure itself is very big. The cases where `__del__` won't be called look acceptable to me.

1. `__del__` never called
    1. the interpreter is shutting down
        1. OS will free memory anyways when the interpreter terminates.
    1. there is a chain of circular references and cyclic garbage collection is disabled and one the objects in the chain has a reference to an instance of WaitSet
        1. It's not just the wait set, all the other objects in the chain are leaked too. The user must fix their circular reference bug if they don't want memory leaks.
1. `__del__` not called for a long time
    1. User messed with garbage collector settings
         1. They've got bigger problems. If this isn't getting collected then a lot of other stuff isn't either
    1.  Something keeping the reference count from hitting zero (ex: An exception was raised and the WaitSet is referenced by a local variable in one of the stack frames in `sys.last_traceback` and another exception hasn't been raised for a while)
        1. Explicit method or not the class instance and member variables will use memory until they are garbage collected (at which point `__del__` will be called unless the interpreter is shutting down). An explicit method can only free the C structure. I'm not sure how big the C structure is; if it's really big then it is worth freeing it sooner.

Reply:
I am pretty sure that without fiddling with any setting the GC won't collect an instance "immediately" but might only do that after minutes. And if the user want's to ensure that the C structure is deleted there needs to be a way to do so (without triggering the GC). Therefore I think it is better to expect that the calling code invokes a cleanup function explicitly. If the caller fails to do so the `__del__` method can this do it for them to be on the safe side.

Reply:
Resolved by adding `WaitSet.destroy()` and having the executor use it via `with` d8521cda33434f5d52c8be8def1c6a40964fdda8